core/src/main/kotlin/endpoints/AgentManager.kt

```diff
@@ -549,10 +533,63 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
         }
     }
 
-    internal suspend fun AgentInfo.commitChanges() {
+    private suspend fun AgentInfo.commitChanges() {
         persistToDatabase()
+        notifyAgents(id)
+    }
+
+    internal suspend fun notifyAgents(agentId: String) {
         notifyAllAgents()
-        notifySingleAgent(id)
+        notifySingleAgent(agentId)
+    }
+
+    private suspend fun wrapInstanceBusy(
+        instanceKey: InstanceKey,
+        block: suspend AgentWsSession.() -> Unit,
+    ): Unit = getInstanceState(instanceKey)?.let { instanceState ->
+        val agentId = instanceKey.agentKey.agentId
+        updateInstanceStatus(instanceKey, AgentStatus.BUSY).also { instance ->
+            if (instance?.all { it.value.status == AgentStatus.BUSY } == true) {
+                notifyAgents(agentId)
+                logger.debug { "Agent $agentId is busy." }
+            }
+            logger.trace { "Instance $instanceKey of agent $agentId is busy." }
+        }
+        try {
+            block(instanceState.agentWsSession)
+        } finally {
+            updateInstanceStatus(instanceKey, AgentStatus.ONLINE).also {
+                notifyAgents(agentId)
+                logger.debug { "Agent $agentId is online." }
+            }
+        }
+    } ?: logger.warn { "Instance $instanceKey is not found" }
+
+    internal fun updateInstanceStatus(
+        instanceKey: InstanceKey,
+        status: AgentStatus,
+    ) = getInstanceState(instanceKey)?.let { instanceState ->
+        val agent = instanceKey.agentKey
+        _instances.updateAndGet { instances ->
+            instances[agent]?.let { instance ->
+                logger.trace { "instance $instanceKey changed status, new status is $status" }
+                instances.put(agent, instance.put(instanceKey, instanceState.copy(status = status)))
+            } ?: instances
+        }[agent]
+    }
+
+    private fun getInstanceState(instanceKey: InstanceKey) = _instances.value[instanceKey.agentKey]?.get(instanceKey)
+
+    fun getStatus(agentId: String): AgentStatus = instanceIds(agentId).let { instances ->
```

Comment:
Can you check this variant of these function 
`fun getStatus(agentId: String): AgentStatus = instanceIds(agentId).takeIf { it.isNotEmpty() }?.let { instances ->
        entryOrNull(agentId)?.agent?.takeIf { it.isRegistered }?.let {
            val statuses = instances.map { it.value.status }.distinct()
            AgentStatus.ONLINE.takeIf { AgentStatus.ONLINE in statuses } ?: AgentStatus.BUSY
        } ?: AgentStatus.NOT_REGISTERED
    } ?: AgentStatus.OFFLINE`
Offline status can be only if agent doesn't have instances.